x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DimensionFieldValueFetcher.java

```diff
@@ -42,12 +42,18 @@ static List<FieldValueFetcher> create(final SearchExecutionContext context, fina
             MappedFieldType fieldType = context.getFieldType(dimension);
             assert fieldType != null : "Unknown dimension field type for dimension field: [" + dimension + "]";
 
-            if (context.fieldExistsInIndex(dimension)) {
+            if (context.fieldExistsInIndex(fieldType.name())) {
                 final IndexFieldData<?> fieldData = context.getForField(fieldType, MappedFieldType.FielddataOperation.SEARCH);
-                final String fieldName = context.isMultiField(dimension)
-                    ? fieldType.name().substring(0, fieldType.name().lastIndexOf('.'))
-                    : fieldType.name();
-                fetchers.add(new DimensionFieldValueFetcher(fieldType, fieldData));
+                if (fieldType instanceof FlattenedFieldMapper.KeyedFlattenedFieldType flattenedFieldType) {
+                    // Name of the field type and name of the dimension are different in this case.
+                    var dimensionName = flattenedFieldType.rootName() + '.' + flattenedFieldType.key();
```

Comment:
Can you come up with a scenario when flattened field can be a multi-field? Cause right now this does not support that.

Reply:
I can't think of a scenario. A flattened field isn't a leaf field, but an object field that gets indexed as if it is a leaf field. So you can indexing it into a single numeric and keyword field via multi fields isn't possible.